arrows/core/derive_metadata.cxx

```diff
@@ -0,0 +1,374 @@
+  try
+  {
+    // Attempt to acquire sensor FOV
+    sensor_fov = get_sensor_fov( metadata );
+  }
+  catch ( kv::invalid_value const& e )
+  {
+    // Fall back to calculating GSD from target width
+    double const target_width = get_target_width( metadata );
+
+    return target_width / frame_width;
+  }
+
+  // Fill in possibly missing sensor FOV dimension
+  if( !sensor_fov[ 0 ] )
+  {
+    sensor_fov[ 0 ] = sensor_fov[ 1 ] * frame_width / frame_height;
+  }
+  else if( !sensor_fov[ 1 ] )
+  {
+    sensor_fov[ 1 ] = sensor_fov[ 0 ] * frame_height / frame_width;
+  }
+
+  // Intermediate Values
+  kv::rotation_d const total_rotation = get_total_rotation( metadata );
+  double yaw, pitch, roll;
+  total_rotation.get_yaw_pitch_roll( yaw, pitch, roll );
+
+  double const slant_range = get_slant_range( metadata );
+  double const altitude_difference = slant_range * sin( -pitch );
+
+  // Approximate dimensions of image on ground plane
+  double const target_width = 2.0 * slant_range *
+                              std::tan( sensor_fov[ 0 ] / 2.0 );
+  double const target_height =
+    2.0 * ( slant_range * std::cos( pitch ) - altitude_difference *
+            std::tan( kv::pi_over_2 - sensor_fov[ 1 ] / 2 ) );
+
+  // GSD is the geometric mean of each dimensions's GSD
+  double const x_gsd = target_width / frame_width;
+  double const y_gsd = target_height / frame_height;
+  double const gsd = std::sqrt( x_gsd * y_gsd );
+  return gsd;
+}
+
+/// \brief Compute Video NIIRS image quality measure
+/// \note  Based on General Image Quality Equation v5 (GIQE5)
+/// \see   https://gwg.nga.mil/ntb/baseline/docs/GIQE-5_for_Public_Release.pdf
+double
+compute_vniirs( double gsd, double rer, double snr )
+{
+  // Taken from Table 2
+  constexpr double a0 = 9.57, a1 = -3.32, a2 = 3.32,
+                   a3 = -1.9, a4 = -2.0,  a5 = -1.8;
+
+  constexpr double meters_to_inches = 1.0 / 0.0254;
+  gsd *= meters_to_inches;
+
+  double const log10_gsd = std::log10( gsd );
+  double const log10_rer = std::log10( rer );
+
+  auto vniirs = a0 +
+                a1 * log10_gsd +
+                a2 * ( 1.0 - std::exp( a3 / snr ) ) * log10_rer +
+                a4 * std::pow( log10_rer, 4 ) +
+                a5 / snr;
+  return vniirs;
+}
+
+/// \brief Compute Relative Edge Response
+/// \todo Implement (should take in image as sole param)
+double
+compute_rer()
+{
+  return 0.3; // Dummy value within reasonable range
+}
+
+/// \brief Compute Signal-to-Noise Ratio
+/// \todo  Implement (should take in image as sole param)
+double
+compute_snr()
+{
+  return 15.0; // Dummy value within reasonable range
+}
+
+} // end namespace
+
+kwiver::vital::metadata_vector
+compute_derived_metadata( kwiver::vital::metadata_vector metadata_vec,
+                          size_t frame_width, size_t frame_height )
+{
+  kv::metadata_vector updated_values;
+
+  for( auto const metadata : metadata_vec )
+  {
+    // Deep copy metadata
+    auto updated_metadata =
```

Comment:
I can't suggest changes on my own PR I guess...but this should be updated to `std::make_shared< kv::metadata >( metadata->deep_copy() )`. My own branch is pretty dirty right now.